experimental/SetPartitions/src/SetPartition.jl

```diff
@@ -0,0 +1,266 @@
+            pop!(ret[2])
+            push!(ret[1], a)
+        end
+    end
+    
+    SetPartition(ret[1], ret[2])
+end
+
+"""
+    composition_loops(p::SetPartition, q::SetPartition)
+
+Return the composition of `p` and `q` as well as the number of removed loops.
+
+# Examples
+```jldoctest
+julia> composition_loops(SetPartition([1, 2], [2, 1]), SetPartition([1], [1, 1]))
+(SetPartition([1], [1, 1]), 0)
+```
+"""
+function composition_loops(p::SetPartition, q::SetPartition)
+    
+    !is_composable(p, q) ? error("format not fitting") : 
+
+    # Work with copies to not change the input partitions
+    p_copy = copy(p)
+
+    # new_ids dictionary stores the new value we need to assign to the partition,
+    # in order to connect new segments
+    vector_q = helper_new_point_values_vector([p_copy.upper_points, p_copy.lower_points], 
+        [copy(q.upper_points), copy(q.lower_points)])
+    new_ids = Dict{Int, Int}()
+    
+    # mapping the second the lower points of the second partition 
+    # to the upper points of the first partition and merge if connection
+    for (i, n) in enumerate(vector_q[2])
+        if !(n in keys(new_ids))
+            new_ids[n] = p.upper_points[i]
+        else
+            if p.upper_points[i] in keys(new_ids) && get(new_ids, n, -1) in keys(new_ids)
+                # Do path compression if we have the case that we need to merge two tree's 
+                # together and the nodes we operate on are not a root or a leaf
+                for ii in [n]
+                    path = [ii]
+                    already_in = Set()
+                    push!(already_in, get(new_ids, ii, -1))
+                    z = get(new_ids, ii, -1)
+                    while z in keys(new_ids)
+                        push!(path, z)
+                        push!(already_in, z)
+                        z = get(new_ids, z, -1)
+                        if z in already_in
+                            break
+                        end
+                    end
+                    push!(path, z)
+                    for nn in path[1:end-1]
+                        new_ids[nn] = path[end]
+                    end
+                end
+                new_ids[get(new_ids, n, -1)] = get(new_ids, p_copy.upper_points[i], -1)
+            else
+                if !(get(new_ids, n, -1) in keys(new_ids))
+                    new_ids[get(new_ids, n, -1)] = p_copy.upper_points[i]
+                else
+                    new_ids[p_copy.upper_points[i]] = get(new_ids, n, 1)
+                end
+            end
+        end
+    end
+    
+    # final path compression
+    for ii in keys(new_ids)
+        path = [ii]
+        already_in = Set()
+        push!(already_in, get(new_ids, ii, -1))
+        z = get(new_ids, ii, -1)
+        while z in keys(new_ids)
+            push!(path, z)
+            push!(already_in, z)
+            z = get(new_ids, z, -1)
+            if z in already_in
+                break
+            end
+        end
```

Comment:
Accessing entries of a dictionary is comparatively expensive. Don't do it again and again when you can get the data once and then keep it around. 

Also, if you create a `Set` via `Set()` this gives a `Set{Any}` which is not type stable, as it has to deal with *arbitrary* content. But here yo already know the type of the content. So you can write `Set{SomeType}()`. Even better: just create the set directly with the initial value, then Julia can infer the correct type.

So, here i a much more idiomatic way to write this code would be like this 
```suggestion
    for (ii, z) in new_ids
        path = [ii]
        already_in = Set((z,))
        while z in keys(new_ids)
            push!(path, z)
            push!(already_in, z)
            z = get(new_ids, z, -1)
            z in already_in && break
        end
```